src/main/java/com/github/elenterius/biomancy/init/AcidInteractions.java

```diff
@@ -157,4 +166,59 @@ else if (livingEntity.tickCount % 10 == 0 && livingEntity.getRandom().nextFloat(
 		}
 	}
 
+	private static final String TIMER_KEY = "biomancy:digestion_timer";
+	private static final float EFFICIENCY = 0.8f;
+	public static void tryDigest(ItemEntity itemEntity, boolean onClient) {
+		DigestingRecipe recipe = getDigestionRecipe(itemEntity);
+		if (!digestible(itemEntity,recipe)) return;
+		CompoundTag data = itemEntity.getPersistentData();
+		if (data.contains(TIMER_KEY) && data.getInt(TIMER_KEY) < 10) {
+			data.putInt(TIMER_KEY, data.getInt(TIMER_KEY) + 1);
+		} else if (data.contains(TIMER_KEY) && data.getInt(TIMER_KEY) >= 10 && !onClient){
+			digestIntoNutrientPasteStacks(itemEntity,recipe);
+			itemEntity.getPersistentData().remove(TIMER_KEY);
+		} else {
+			data.putInt(TIMER_KEY,1);
+		}
+	}
+
+	private static void digestIntoNutrientPasteStacks(ItemEntity itemEntity, DigestingRecipe recipe) {
+		BehavioralInventory<?> tempInventory = BehavioralInventory.createServerContents(1,player->false,()->{});
+		tempInventory.insertItemStack(itemEntity.getItem());
+		ItemStack resultStack = recipe.assemble(tempInventory,itemEntity.level().registryAccess());
+		int totalToOutput = (int)Math.floor(resultStack.getCount()*itemEntity.getItem().getCount()*EFFICIENCY);
+		if (totalToOutput > 64) totalToOutput = splitIntoStacks(itemEntity,totalToOutput);
+		itemEntity.setItem(new ItemStack(ModItems.NUTRIENT_PASTE.get(), totalToOutput));
+		itemEntity.playSound(SoundEvents.PLAYER_BURP);
+	}
+
+	private static int splitIntoStacks(ItemEntity itemEntity, int numToSplit) {
+		Level level = itemEntity.level();
+		while (numToSplit > 64) {
+			DefaultDispenseItemBehavior.spawnItem(level,new ItemStack(ModItems.NUTRIENT_PASTE.get(),64),1, Direction.UP, itemEntity.position());
+			numToSplit -= 64;
+		}
+		return numToSplit;
+	}
+
+
+	@SuppressWarnings("RedundantIfStatement")
+	private static boolean digestible(ItemEntity itemEntity, @Nullable DigestingRecipe recipe) {
+		if (recipe == null) return false;
+		if (!itemEntity.isInFluidType(ModFluids.ACID_TYPE.get()) && !itemEntity.level().getBlockState(itemEntity.blockPosition()).is(ModBlocks.ACID_CAULDRON.get())) return false;
+		//Inside method to prevent missing registry object errors during init
+		Item[] blacklistedItems = {ModItems.NUTRIENT_PASTE.get(),ModItems.NUTRIENT_BAR.get(),ModItems.LIVING_FLESH.get()};
+		if (ArrayUtils.contains(blacklistedItems,itemEntity.getItem().getItem())) return false;
+		return true;
+	}
+
+	private static DigestingRecipe cachedRecipe = null;
```

Comment:
i should have been more clear what i mean with caching. I meant storing the recipe Id in the persistent data.

Example:
```java
	public static void tryDigest(ItemEntity itemEntity) {
		ItemStack stack = itemEntity.getItem();
		CompoundTag tag = getOrCreateTag(itemEntity.getPersistentData(), DIGESTION_KEY);

		ResourceLocation lastRecipeId = ResourceLocation.tryParse(tag.getString(LAST_RECIPE_KEY));
		Optional<Pair<ResourceLocation, DigestingRecipe>> optionalPair = DigesterBlockEntity.RECIPE_TYPE.get().getRecipeForIngredient(itemEntity.level(), stack, lastRecipeId);

		if (optionalPair.isEmpty()) return;

		ResourceLocation recipeId = optionalPair.get().getFirst();
		DigestingRecipe recipe = optionalPair.get().getSecond();

		tag.putString(LAST_RECIPE_KEY, recipeId.toString());

		int timer = tag.getInt(TIMER_KEY);
		if (timer < 10) {
			tag.putInt(TIMER_KEY, timer + 1);
		}
		else {
			if (!itemEntity.level().isClientSide) {
				RecipeWrapper inventory = new RecipeWrapper(new ItemStackHandler(1));
				inventory.setItem(0, stack);
				digestIntoNutrientPasteStacks(itemEntity, inventory, recipe);
			}
			tag.remove(TIMER_KEY);
		}
	}
```

Reply:
Ah ok, I'll do that.
Would that not still need to query the recipe every time it is run, though? Saying as you're not storing a reference to the recipe itself?